Add log encryption

jkmassel opened the pull request:
Implements log encryption.

Replaces the former implementation (which used GPLv3), which we're not able to use due to license incompatibility. 

**To test:**
- Ensure CI passes
- Manually run the `androidTest/util/EncyptionUtilsTest` test suite, which covers this code.

**To review:**
I've added comments in a few places for discussion or validation.

**PR submission checklist:**
- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jkmassel commented on WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.kt:
```diff
@@ -0,0 +1,91 @@
+        val log = logWithContent(testLogString)
+
+        val json = JSONObject(log.readText())
+        assertEquals(
```
These are all spread out because they won't fit on one line and the linter didn't like that. 

IMHO this is still quite readable, but LMK if not!

jkmassel commented on WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedLogWriter.kt:
```diff
@@ -0,0 +1,95 @@
+     */
+    fun write(string: String) {
+        val encryptedString = encryptMessage(string, SecretStream.TAG_MESSAGE)
+        fileWriter.write("\t\t\"$encryptedString\",\n")
```
This implementation uses the [same streaming write approach as on iOS](https://github.com/Automattic/Automattic-Tracks-iOS/pull/120), because it can be much more memory-efficient without much more complexity. 

The manually-create-the-json stuff is covered by a test, so IMHO the little bit of added complexity is worth it for the efficiency, but it was discussed in that PR, so I figured I'd mention it in this one as well.

oguzkocer replied:
I am totally good with this approach for this specific usecase 👍

jkmassel commented on WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedLogReader.kt:
```diff
@@ -0,0 +1,80 @@
+}
+
+// On Android base64 has lots of options, so define an extension to make it easier to avoid decoding issues.
+private fun String.base64Decode(): ByteArray {
```
I'm not sure this is idiomatic from a project standpoint? Having the extension written like this makes the implementation far clearer (`json.getString("encryptedKey").base64Decode()` seems slightly nicer to read than `base64Decode(json.getString("encryptedKey"))` in terms of understanding what is going on (we're getting the json, then we're transforming it).

The extension is `private`, so should be scoped to this file. 

The alternative is just doing something like https://github.com/wordpress-mobile/WordPress-Android/pull/11562/files#diff-66f71d6c155d5b8d978f5a102c856f98R92-R94, but in that case it didn't seem to affect readability at all.

Open to changing this if it isn't considered "right" :)

oguzkocer replied:
I think the usage you have is totally fine :)

jkmassel commented on WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedLogWriter.kt:
```diff
@@ -0,0 +1,95 @@
+        check(SecretStream.Checker.headerCheck(header.size))
+
+        val encodedEncryptedKey = base64Encode(key.bytes)
+        check(encodedEncryptedKey.length == 108) {
```
This is a somewhat "magic" number – it's the right number for this implementation, and is enforced by a unit test, but having the check here helps debug implementation issues (for instance, if a developer tries to use the wrong kind of `KeyPair` with this class).

It's a constant constant – ie, this will never change as long as this implementation exists, so it shouldn't be extracted to a class constant – if a developer changes this value, they break the compatibility with the server-side decryption logic, which is why it's directly in the code like this.

oguzkocer replied:
If there is ever a time to use numbers directly, I think this would be it. I really like the `checks` all over and the message for that already explains the "magic" bit. (kind of) 👍

jkmassel commented on WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedLogWriter.kt:
```diff
@@ -0,0 +1,95 @@
+        }
+
+        val encodedHeader = base64Encode(header)
+        check(encodedHeader.length == 32) {
```
Same as above.

jkmassel commented:
@oguzkocer – I've addressed the feedback we discussed:
- [x] Using constants for JSON parsing keys
- [x] Auditing for `require` vs `check` everywhere
- [x] Using immutable values even when passing by reference

This is ready for another look, and I _think_ should be ready to merge!

Thank you!!!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedSecretStreamKey.kt:
```diff
@@ -0,0 +1,54 @@
+        /**
+         * The expected size (in bytes) of an Encrypted Secret Stream Key.
+         */
+        val size: Int = SecretStream.KEYBYTES + Box.SEALBYTES
```
**AS suggesion:** We can make this a `const`.

```suggestion
        const val size: Int = SecretStream.KEYBYTES + Box.SEALBYTES
```
